src/python/pants/backend/kotlin/lint/ktlint/rules.py

```diff
@@ -0,0 +1,110 @@
+# Copyright 2022 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+import logging
+from dataclasses import dataclass
+
+from pants.backend.kotlin.lint.ktlint.skip_field import SkipKtlintField
+from pants.backend.kotlin.lint.ktlint.subsystem import KtlintSubsystem
+from pants.backend.kotlin.target_types import KotlinSourceField
+from pants.core.goals.fmt import FmtRequest, FmtResult
+from pants.core.goals.generate_lockfiles import GenerateToolLockfileSentinel
+from pants.engine.fs import Digest
+from pants.engine.internals.native_engine import Snapshot
+from pants.engine.internals.selectors import Get
+from pants.engine.process import ProcessResult
+from pants.engine.rules import collect_rules, rule
+from pants.engine.target import FieldSet, Target
+from pants.engine.unions import UnionRule
+from pants.jvm.jdk_rules import InternalJdk, JvmProcess
+from pants.jvm.resolve import jvm_tool
+from pants.jvm.resolve.coursier_fetch import ToolClasspath, ToolClasspathRequest
+from pants.jvm.resolve.jvm_tool import GenerateJvmLockfileFromTool
+from pants.util.logging import LogLevel
+from pants.util.strutil import pluralize
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass(frozen=True)
+class KtlintFieldSet(FieldSet):
+    required_fields = (KotlinSourceField,)
+
+    source: KotlinSourceField
+
+    @classmethod
+    def opt_out(cls, tgt: Target) -> bool:
+        return tgt.get(SkipKtlintField).value
+
+
+class KtlintRequest(FmtRequest):
+    field_set_type = KtlintFieldSet
+    name = KtlintSubsystem.options_scope
+
+
+class KtlintToolLockfileSentinel(GenerateToolLockfileSentinel):
+    resolve_name = KtlintSubsystem.options_scope
+
+
+@dataclass(frozen=True)
+class Setup:
+    process: JvmProcess
+
+
+@rule(level=LogLevel.DEBUG)
+async def setup_ktlint(
+    request: KtlintRequest,
+    jdk: InternalJdk,
+) -> Setup:
```

Comment:
The redirection through the `Setup` type/`@rule` should no longer be necessary, since @thejcannon's recent change means there is only one caller (and since the `Request` object holds a field set, the `Setup` will still end up memoized per-caller anyway).

Reply:
I had copied much of this code from google_java_format. Have any linters / formatters been ported to this new way of doing things (and which I could use as an example)?

Reply:
docformatter. But even that has more indirection than necessary. There is no need for a helper rule.

Reply:
I can purge the other formatters sometime this week when I'm doing my routine Pants code gardening :tulip:

Reply:
I got rid of the Setup type and rule in this PR. Much simpler now, thanks for good ideas!